execute/plugin.go

```diff
@@ -537,9 +544,98 @@ func (p *Plugin) validateReport(
+		// been executed, so we don't want to re-execute them.
+		// This gives the exec plugin a chance to remedy the situation
+		// by selecting a different set of messages.
+		return false, decodedReport, nil
+	}
+	if err != nil {
+		// TODO: should we return true here if we couldn't check for already executed messages?
+		return false, decodedReport, fmt.Errorf("checking for already executed messages failed: %w", err)
+	}
+
 	return true, decodedReport, nil
 }
 
+// checkAlreadyExecuted checks if the messages in the report have already been executed
+// on the destination chain. It queries the DB for executed messages in the given sequence
+// number range for each source chain in the report.
+func (p *Plugin) checkAlreadyExecuted(
+	ctx context.Context,
+	lggr logger.Logger,
+	seqNrRangesBySource map[cciptypes.ChainSelector]snRangeSetPair,
+) error {
+	// TODO: batch these queries? these are all DB reads.
+	// maybe some alternative queries exist.
+	for sourceChainSelector, seqNrRange := range seqNrRangesBySource {
+		executed, err := p.ccipReader.ExecutedMessages(ctx, sourceChainSelector, seqNrRange.snRange)
+		if err != nil {
+			return fmt.Errorf("couldn't check if messages already executed: %w", err)
+		}
+
+		if intersection := mapset.NewSet(executed...).Intersect(seqNrRange.set); !intersection.IsEmpty() {
+			// Some messages in the report have been executed, don't accept it.
+			reportSeqNrsSlice := seqNrRange.set.ToSlice()
+			lggr.Warnw("some messages in report already executed",
+				"alreadyExecuted", executed,
+				"reportSeqNrs", reportSeqNrsSlice,
+			)
+			return fmt.Errorf("%w: already executed messages %+v report seq nrs %+v",
+				errAlreadyExecuted, executed, reportSeqNrsSlice)
+		}
+	}
+
+	return nil
+}
+
+// snRangeSetPair is an internal data structure used to store a sequence number range
+// and a set of sequence numbers simultaneously.
+type snRangeSetPair struct {
+	// snRange is a range of [min, max] sequence numbers of the messages in the report for a particular source chain.
+	// it is used to query the CCIPReader for executed messages.
+	snRange cciptypes.SeqNumRange
+	// set is the sequence numbers of the messages in the report for a particular source chain.
+	// it is used to check whether the returned array from CCIPReader has a non-empty intersection
+	// with the set of sequence numbers in the report.
+	set mapset.Set[cciptypes.SeqNum]
+}
+
+// getSnRangeSetPairsBySource returns a map of source chain selector to
+// the sequence number range of the messages in the report.
+func getSnRangeSetPairsBySource(
+	chainReports []cciptypes.ExecutePluginReportSingleChain,
+) map[cciptypes.ChainSelector]snRangeSetPair {
+	seqNrRangesBySource := make(map[cciptypes.ChainSelector]snRangeSetPair)
+	for _, chainReport := range chainReports {
+		cmpr := func(a, b cciptypes.Message) int {
+			return cmp.Compare(a.Header.SequenceNumber, b.Header.SequenceNumber)
+		}
+		minMsg := slices.MinFunc(chainReport.Messages, cmpr)
```

Comment:
MinFunc/MaxFunc gonna panic when a slice is empty. Probably either using a different method or just adding length check should do the job. I recommend adding a testcase for it

Reply:
Oh I guess that makes sense with this API.

Reply:
Will add a check for message length in the chain report. We check that ChainReport is > 0 length and report is non-nil, I guess its a super-bug if the messages in it are nil, but will add the check anyway to be defensive.